go-controller/pkg/util/kube.go

```diff
@@ -719,51 +733,88 @@ func NoHostSubnet(node *v1.Node) bool {
-// Function fn takes a bool pointer "shortcut" that, when set to true inside fn, ends the iteration; this is useful when fn
-// checks for a condition on the endpoints and we want to return as soon as the condition is satisfied.
-func ForEachEligibleEndpoint(endpointSlice *discovery.EndpointSlice, service *kapi.Service, fn func(ep discovery.Endpoint, shortcut *bool)) {
+// GetEligibleEndpoints takes a list of endpoints (provided from endpoint slices) and a pointer to a service
+// and returns a list of eligible endpoints. The endpoint selection implements KEP-1669
+// (https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/1669-proxy-terminating-endpoints/README.md)
+// First it selects only ready endpoints, which effectively filters out all terminating endpoints
+// (an endpoint that is terminating shows ready=false). Then, only if no ready endpoints are found,
+// it falls back to serving terminating endpoints.
+// However, if the service has Spec.PublishNotReadyAddresses set, all provided endpoints are always returned.
+// PublishNotReadyAddresses tells endpoint consumers to disregard any indications of ready/not-ready and is generally
+// used together with headless services so that DNS records of all endpoints (ready or not) are always published.
+func GetEligibleEndpoints(endpoints []discovery.Endpoint, service *kapi.Service) []discovery.Endpoint {
 	includeTerminating := service != nil && service.Spec.PublishNotReadyAddresses
-	var shortcut bool
-	for _, endpoint := range endpointSlice.Endpoints {
-		if IsEndpointEligible(endpoint, includeTerminating) {
-			fn(endpoint, &shortcut)
-			if shortcut {
-				// shortcircuit the whole iteration
-				return
+	var eligibleEndpoints []discovery.Endpoint
+
+	for _, endpoint := range endpoints {
+		if includeTerminating || IsEndpointReady(endpoint) {
+			eligibleEndpoints = append(eligibleEndpoints, endpoint)
+		}
+	}
+	serviceStr := ""
+	if service != nil {
+		serviceStr = fmt.Sprintf(" (service %s/%s)", service.Namespace, service.Name)
+	}
+	klog.V(5).Infof("Endpoint selection%s: found %d ready endpoints", serviceStr, len(eligibleEndpoints))
+
+	// Fallback to serving terminating endpoints (ready=false, serving=true, terminating=true) only if none are ready.
+	if len(eligibleEndpoints) == 0 {
+		for _, endpoint := range endpoints {
+			if IsEndpointServing(endpoint) && IsEndpointTerminating(endpoint) {
+				eligibleEndpoints = append(eligibleEndpoints, endpoint)
 			}
 		}
+		klog.V(5).Infof("Endpoint selection%s: fallback to %d serving & terminating endpoints",
+			serviceStr, len(eligibleEndpoints))
+	}
+
+	return eligibleEndpoints
+}
+
+// ForEachEligibleEndpoint first finds eligible endpoints from a provided list of endpoints
+// and then iterates through them to apply the given function fn to each eligible endpoint.
+func ForEachEligibleEndpoint(endpoints []discovery.Endpoint, service *kapi.Service, fn func(ep discovery.Endpoint)) {
```

Comment:
I have the feeling that if you implemented this function like
```
// forEachSelectedAndEligibleEndpoint filters the endpoints by the given condition, 
// then selects the elegible ones and a applies the provided function to them. 
// Eligible endpoints are serving & ready endpoints. If there are none, eligible endpoints
// are serving & terminating endpoints.
// Returns true when elegible endpoints where ready endpoints or false if they were 
// terminating endpoints
func forEachSelectedAndEligibleEndpoint(endpoints []discovery.Endpoint, condFn func(ep discovery.Endpoint) bool, fn func(ep discovery.Endpoint)) bool
```
then all the other logic could be written in terms of this function in a much simpler way. Note the un-exported nature of the method.

The implementation of this method would just accumulate selected terminating and non terminating endpoints in separate slices and then iterate in either depending non terminating slice being empty or not.

You will probably need to translate `[]discovery.EndpointSlice` to  `[]discovery.Endpoint` in some places so you could have an utility method for that as well.

Reply:
Done. Does that look like what you had in mind? I just didn't create two separate slices for ready and serving&terminating endpoints because the condition function, which we use to check whether an endpoint is local, needs to be applied first. Only afterwards we make a decision based on the endpoint readiness, as I explained to Jacob in my other comment.
Thanks!

Reply:
I also renamed `GetLocalEndpointAddresses` as `GetLocalEligibleEndpointAddresses` to make it clear that it runs the endpoint selection logic as well.